### ISSUE TITLE
HTML: passwordrules attribute

### ISSUE THREAD

**Issue by annevk** [mozilla/standards-positions#61]
In https://github.com/whatwg/html/issues/3518 Apple engineers propose a new attribute for `<input type=password>` that'll assist password generators doing the right thing.

I'm personally enthusiastic about the idea.

(There's many small incremental features like this added to WHATWG standards. It's not quite clear to me if we want an issue for each of them, but we probably do want a Mozilla position on each of them, so... Filing this one as a test.)

**Comment by jonathanKingston** [comment]
> "Restrictions on passwords beyond minimum length (and maybe a large maximum length) are all fundamentally bad" ... Do we really want to be adding a feature whose primary use-case is making it easier for already-broken sites to continue being broken?

Cementing this into a standard would somewhat agree with the status quo, I'm with Tab on this.

I had a [proposal](https://github.com/w3c/webappsec/issues/250#issuecomment-136005063) that merely permitted a site selecting the entropy which would restrict the length and also inform the password manager how long it should be safe to store. For example most sites aren't going to want a 50mb password but also a 50 character one might be overkill for a voucher token.

**Comment by linuxwolf** [comment]
On the one hand, I think it would be very helpful to get hints for what a site wants for passwords.  It helps password generators come up with something for the user faster.

However, I think this proposal is moving in the wrong direction.  As @jonathanKingston points out, it cements – even encourages – bad practices, without promoting good practices (e.g., minimum length).

My personal opinion would be we lean toward `harmful` on this.

**Comment by annevk** [comment]
I see some positives for sites with hard-to-change backends not thwarting password managers, but perhaps that's best catalogued in https://github.com/apple/password-manager-resources and not the frontend of those sites.

@linuxwolf are you interested in writing a PR for this, as nobody contested your comment and the one above?

**Comment by mnoorenberghe** [comment]
For the most part I agree with https://github.com/whatwg/html/issues/3518#issuecomment-419470439 from Chromium but I also have some other concerns:
* While I find the use of `<key>`:`<value>;` more clear, it differs from syntax of CSP and feature policy. Would this be the first attribute that uses this syntax? Is it worth diverging from that prior art? @annevk had similar comments at https://github.com/whatwg/html/issues/3518#issuecomment-369875147
* The overlap with `setCustomValidity` and @pattern: https://github.com/whatwg/html/issues/3518#issuecomment-370074201
* This is likely to have low adoption and may encourage sites to add password requirements (as mentioned above too): https://github.com/whatwg/html/issues/3518#issuecomment-370249052

Without even taking into account `ValidityState`, our telemetry data shows that only 4% of generated passwords are edited after generation. That seems okay and I suspect only a small fraction of sites which rejected these generated passwords would even know to adopt this new attribute. (This telemetry mostly catches when sites do client-side validation before submission so the user knows to edit and may slightly under-count rejected passwords that are only caught on the server side.)

https://github.com/whatwg/html/issues/3518#issuecomment-370253483 is the approach we were planning to take after supporting `minlength`/`maxlength`/`pattern` attributes (and possibly checking `ValidityState.customError`) so we don't have any plans to adopt this new `passwordrules` attribute. We will also use a list of recipes to override the generated password format for sites which don't use the constraint validation attributes but have unique requirements. 

IMO this new attribute would be harmful if it encourages new password restrictions or if authors use it instead of `minlength`/`maxlength`/`pattern`/`setCustomValidity`, which are also useful outside of password generators.

**Comment by jrmuizel** [comment]
I'm not convinced that supporting the password rules attribute will actually encourage sites to add new password restrictions. It's hard to imagine a scenario where a develop wants to add restrictions but the developer cost for enforcing the restrictions is the bottleneck.

I think the potential user value during password generation and the ease of use by sites outweighs any potential downside here.

**Comment by zcorpan** [comment]
In 2023 the attribute had very low adoption: https://github.com/whatwg/html/issues/3518#issuecomment-1623940647

It's possible it would be adopted more if there was cross-browser support and better docs. Some browsers have out-of-band site-specific password rules (e.g. https://github.com/apple/password-manager-resources/blob/main/quirks/password-rules.json ), which seems likely to continue to exist as long as websites have password restrictions and don't use the `passwordrules` attribute. But it seems reasonable to allow websites to specify their own rules to improve UX, even if adoption continues to be low.

**Comment by jrmuizel** [comment]
My suggestion is that our position be positive.

**Comment by issammani** [comment]
I agree with with @jrmuizel. I think the benefits outweigh any potential downsides. As @zcorpan stated with better docs and better support we can get some adoption going, although I think we will always need the pre-defined list as a fallback.

**Comment by Nettsentrisk** [comment]
Working on implementing support for password managers to be able to generate a password in line with the specified requirements, for example a password between 14 to 64 characters, containing at least three of the following types of characters: uppercase, lowercase, digit, special characters.

I guess it would be technically possible to specify a regex in a pattern attribute that mandates this, but would such a regex be technically feasible for a browser/plugin to use for generating passwords in line with that rule?

The utility of the proposed passwordrules attribute is that it makes it more transparent for the HTML developer what the rules are, rather than using a regex pattern.

What is missing from the passwordrules proposal is specifying a type of rule I had as an example here, namely requiring X of Y character classes; the proposal as it stands only allows you to specify which character classes to require, period.

This would be the pattern attribute to enforce the example rule:
```^(?=(?:.*[a-z]){0,})(?=(?:.*[A-Z]){0,})(?=(?:.*\d){0,})(?=(?:.*[\W_]){0,})(?=.{14,64})(?:(?=.*[a-z])(?=.*[A-Z])(?=.*\d)|(?=.*[a-z])(?=.*[A-Z])(?=.*[\W_])|(?=.*[a-z])(?=.*\d)(?=.*[\W_])|(?=.*[A-Z])(?=.*\d)(?=.*[\W_])).*$```

Whereas a passwordrules attribute could be:
```required: (upper, lower, digit, special){3,}```

To me the latter would drive more adoption and be simpler for the developer to read, update, and implement. I would think it would be simpler for password generators (in browser or extension/plugin) to use the latter scheme instead of having to parse regex patterns for operative rules to use.

**Comment by denyshon** [comment]
Since the position has changed, could you please consider [updating it in the file](https://github.com/mozilla/standards-positions/blob/6763638ac34c3e5b19c19a9a49f3d78f3c71cee1/activities.yml#L2144-L2154)?
